x-pack/plugins/timelines/public/components/t_grid/toolbar/fields_browser/index.tsx

```diff
@@ -67,14 +69,23 @@ export const StatefulFieldsBrowserComponent: React.FC<FieldBrowserProps> = ({
     };
   }, [debouncedApplyFilterInput]);
 
+  const selectionFilteredBrowserFields = useMemo<BrowserFields>(
+    () =>
+      filterSelectedEnabled
+        ? filterSelectedBrowserFields({ browserFields, columnHeaders })
+        : browserFields,
+    [browserFields, columnHeaders, filterSelectedEnabled]
+  );
```

Comment:
filter by selected fields first if needed, and keep the result memoized for the search filter to be applied on top of it.